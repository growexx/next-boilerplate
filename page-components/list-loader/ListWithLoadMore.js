import React, { useEffect, useState } from 'react'
import { List, Avatar, Button, Skeleton } from 'antd'

import request from '@shared/utils/request'
import { API_ENDPOINTS } from '@shared/constants'

const count = 3

const ListWithLoadMore = () => {
  const [list, setList] = useState([])
  const [loading, setLoading] = useState(false)
  const [initLoading, setInitLoading] = useState(true)
  const [data, setData] = useState([])

  useEffect(() => {
    getData((res) => {
      setInitLoading(false)
      setData(res.results)
      setList(res.results)
    })
  }, [])

  const getData = (callback) => {
    request(API_ENDPOINTS.LIST, {
      method: 'GET'
    }).then((res) => {
      callback(res)
    })
  }

  const onLoadMore = () => {
    setLoading(true)
    setList(data.concat([...new Array(count)].map(() => ({ loading: true, name: {} }))))
    getData((res) => {
      const listData = data.concat(res.results)
      setLoading(false)
      setList(listData)
      setData(listData, () => {
        // Resetting window's offsetTop so as to display react-virtualized demo underfloor.
        // In real scene, you can using public method of react-virtualized:
        // https://stackoverflow.com/questions/46700726/how-to-use-public-method-updateposition-of-react-virtualized
        // window.dispatchEvent(new Event('resize'))
      })
    })
  }

  const loadMore =
    !initLoading && !loading ? (
      <div
        style={{
          textAlign: 'center',
          marginTop: 12,
          height: 32,
          lineHeight: '32px'
        }}
      >
        <Button aria-label="load-more" onClick={onLoadMore}>
          load more
        </Button>
      </div>
    ) : null

  return (
    <List
      className="demo-load-more-list"
      itemLayout="horizontal"
      loadMore={loadMore}
      dataSource={list}
      renderItem={(item) => (
        <List.Item>
          <Skeleton avatar title={false} loading={item.loading} active>
            <List.Item.Meta avatar={<Avatar src={API_ENDPOINTS.LIST_AVATAR} />} title={item.name.last} description={item.email} />
          </Skeleton>
        </List.Item>
      )}
    />
  )
}

export default ListWithLoadMore
