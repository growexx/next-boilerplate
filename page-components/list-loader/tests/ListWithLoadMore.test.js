import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'

import request from '@shared/utils/request'
import ListWithLoadMore from '../ListWithLoadMore'

jest.mock('@shared/utils/request')

describe('<ListWithLoadMore />', () => {
  const promise = Promise.resolve({
    status: 1,
    results: [
      {
        email: 'test@example.com',
        gender: 'male',
        name: {
          first: 'test',
          last: 'test',
          title: 'test'
        }
      }
    ]
  })

  it('should render and match the snapshot', async () => {
    request.mockImplementation(() => promise)
    const {
      container: { firstChild }
    } = render(<ListWithLoadMore />)
    await act(() => promise)
    expect(firstChild).toMatchSnapshot()
  })

  it('call Load More', async () => {
    request.mockImplementation(() => promise)
    render(<ListWithLoadMore />)
    await waitFor(() => expect(request).toHaveBeenCalledTimes(2))
    await act(() => promise)
    fireEvent.click(screen.getByRole('button', { name: 'load-more' }))
    await waitFor(() => expect(request).toHaveBeenCalled())
  })
})
