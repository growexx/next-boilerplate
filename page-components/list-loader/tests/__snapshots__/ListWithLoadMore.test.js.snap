// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListWithLoadMore /> should render and match the snapshot 1`] = `
<div
  class="ant-list ant-list-split ant-list-something-after-last-item demo-load-more-list css-dev-only-do-not-override-10ed4xt"
>
  <div
    class="ant-spin-nested-loading css-dev-only-do-not-override-10ed4xt"
  >
    <div
      class="ant-spin-container"
    >
      <ul
        class="ant-list-items"
      >
        <li
          class="ant-list-item"
        >
          <div
            class="ant-list-item-meta"
          >
            <div
              class="ant-list-item-meta-avatar"
            >
              <span
                class="ant-avatar ant-avatar-circle ant-avatar-image css-dev-only-do-not-override-10ed4xt"
              >
                <img
                  src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png"
                />
              </span>
            </div>
            <div
              class="ant-list-item-meta-content"
            >
              <h4
                class="ant-list-item-meta-title"
              >
                test
              </h4>
              <div
                class="ant-list-item-meta-description"
              >
                test@example.com
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div
    style="text-align: center; margin-top: 12px; height: 32px; line-height: 32px;"
  >
    <button
      aria-label="load-more"
      class="ant-btn css-dev-only-do-not-override-10ed4xt ant-btn-default"
      type="button"
    >
      <span>
        load more
      </span>
    </button>
  </div>
</div>
`;
